File: BfsAndDfs/P494TargetSum/TargetSum.java
package BfsAndDfs.P494TargetSum;

/**
 * @CLassName TargetSum
 * @Description
 * You are given a list of non-negative integers, a1, a2, ..., an, and a target, S. Now you have 2 symbols + and -. For each integer, you should choose one from + and - as its new symbol.
 *
 * Find out how many ways to assign symbols to make sum of integers equal to target S.
 *
 * Example 1:
 *
 * Input: nums is [1, 1, 1, 1, 1], S is 3.
 * Output: 5
 * Explanation:
 *
 * -1+1+1+1+1 = 3
 * +1-1+1+1+1 = 3
 * +1+1-1+1+1 = 3
 * +1+1+1-1+1 = 3
 * +1+1+1+1-1 = 3
 *
 * There are 5 ways to assign symbols to make the sum of nums be target 3.
 * @Author cst
 * @Date 1/21/21  1:27 PM
 */
public class TargetSum {
    private int res = 0;
    public int findTargetSumWays(int[] nums, int S) {
        if (nums == null || nums.length == 0) {
            return res;
        }
        helper(nums, S, 0, 0);
        return res;
    }

    public void helper(int[] nums, int target, int idx, int curSum) {
        if (idx == nums.length) {
            if (target == curSum) {
                res += 1;
            }
            return;
        }

        helper(nums, target, idx + 1, curSum + nums[idx]);
        helper(nums, target, idx + 1, curSum - nums[idx]);
    }
}

File: BfsAndDfs/P494TargetSum/TargetSum.py
import unittest
from typing import List


        
class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        # dfs
        self.memo = {}
        def dfs(nums, idx, target):
            if idx == len(nums):
                return 1 if target == 0 else 0
            if (idx, target) in self.memo:
                return self.memo[(idx, target)]
            res = dfs(nums, idx + 1, target - nums[idx]) + dfs(nums, idx + 1, target + nums[idx])
            self.memo[(idx, target)] = res
            return res
        return dfs(nums, 0, S)

    
class UnitTest(unittest.TestCase):
    def test(self):
        nums = [1, 1, 1, 1, 1]
        S = 3
        self.assertEqual(5, Solution().findTargetSumWays(nums, S))
if __name__ == "__main__":
    unittest.main()